src/app.js

```diff
@@ -18,33 +37,57 @@
 
   /**
    * Create a HTTP server
-   * @param {string} subdomain Subdomain which must be used for authentication
+   * @param {object} app
+   * @param {string} app.subdomain Subdomain which must be used for authentication
+   * @param {number} app.port Port on which the server must be started
+   * @param {boolean} app.directoryListing Whether to have directory listings
    * @param {string} rootDirectory Directory from which to serve the files
-   * @param {number} port Port on which the server must be started
    * @param {boolean} allowExternalPorts If true, the server will only serve files on localhost. Else the files can be accessed from all computers on the network.
    */
-  function app(subdomain, rootDirectory, port, allowExternalPorts) {
+  function app(app, rootDirectory, allowExternalPorts) {
+    var port = app.port;
+
     let httpApp = express();
 
-    function auth(req, res, next) {
-      let user = basicAuth(req);
+    let getLatestService = new getLatest(rootDirectory);
+    httpApp.use(getLatestService.getLatestMiddleware.bind(getLatestService));
+
+    httpApp.use('/resources', express.static(path.join(__dirname + '/resources'), {dotfiles: 'allow'}));
 
-      if (!user || !credentials.check(subdomain, user.name, user.pass)) {
-        res.set('WWW-Authenticate', 'Basic realm=Authorization Required');
-        return res.sendStatus(401);
-      } else {
-        return next();
-      }
+    if (app.auth) {
+      let githubAuthService = new githubAuth(app.auth, app.domain);
+
+      httpApp.use(expressSession(sessionOptions));
+
+      httpApp.use('/callback', githubAuthService.authMiddleware.bind(githubAuthService));
+      httpApp.use(function checkPathSafety(req, res, next) {
+        // if realPath is different from rootDirectory + originalUrl we don't continue
+
+        rp.realpath(rootDirectory + req.originalUrl, function (err, realPath) {
+          realPath = realPath && realPath.replace(/\/$/, '');
+          let originalPath = (rootDirectory + req.originalUrl);
+          originalPath = originalPath && originalPath.replace(/\/$/, '');
+          if (realPath === originalPath) {
+            next();
+          } else {
+            res.status('400').send('Symlinks are disabled');
+          }
+        });
+      });
+      httpApp.use(app.auth.urlPrefix, githubAuthService.authMiddleware.bind(githubAuthService));
+    }
+
+    httpApp.use(express.static(rootDirectory, {dotfiles: 'allow'}));
```

Comment:
Why allow dotfiles?

If this is for directory listing, let's `'deny'` when directory listing is disabled.